src/sync.c

```diff
@@ -276,11 +287,25 @@ void sync_process(sync_t *st, float complex *buffer)
     // if we are still synchronized
     if (st->ready)
     {
+        float samperr = 0, angle = 0;
         for (i = 0; i < partitions_per_band * 19; i += 19)
         {
             adjust_data(buffer, st->phases, LB_START + i, LB_START + i + 19);
             adjust_data(buffer, st->phases, UB_END - i - 19, UB_END - i);
+
+            samperr += phase_diff(st->phases[(LB_START + i) * BLKSZ], st->phases[(LB_START + i + 19) * BLKSZ]);
+            samperr += phase_diff(st->phases[(UB_END - i - 19) * BLKSZ], st->phases[(UB_END - i) * BLKSZ]);
+        }
+        samperr = samperr / (partitions_per_band * 2) * 2048 / 19 / (2 * M_PI);
+        st->samperr = roundf(samperr);
+
+        for (i = 0; i < partitions_per_band * 19 + 1; i += 19)
+        {
+            angle += prev_slope[LB_START + i];
+            angle += prev_slope[UB_END - i];
         }
+        angle /= (partitions_per_band + 1) * 2;
+        st->angle = angle;
```

Comment:
Frequency error causes a linear increase in the phase of each carrier over time, which we're already measuring as `slope`. The average slope gives a very good estimate of the frequency error.